### PR TITLE
Aggregate traintuples/algos

### PR DESCRIPTION
**Pull request by AurelienGasser**
See also: [chaincode "composite" PR](https://github.com/SubstraFoundation/substra-chaincode/pull/5)

- [x] Creation of a new structure used for “inputting” a new aggtuple
- [x] Creation of a new structure corresponding to elements stored in the ledger for a given aggtuple
- [x] Creation of a new structure used for “outputting” an aggtuple; with the following elements
- [x] Creation of a smart contract to create an aggtuple with the input mentioned above.
- [x] Creation of smart contracts to query an aggtuple and query all aggtuples
- [x] Creation of smart contracts similar to logStartTrain, logFailTrain, logSuccesTrain for aggtuples.
- [x] Create a smart contract to add/get/list aggregateAlgo and compositeAlgo, which have the same structure as a normal Algo (and the same input and output)

### CODE REVIEW

**Comment by AurelienGasser** [comment]
There are still a lot of minor "TODO"s, but the main functionality works.

**Comment by AurelienGasser** [comment]
Marking the PR as "Ready for review" just so that we have Travis run unit tests. **Edit**: That didn't work :(

**Comment by AurelienGasser** [comment]
This is ready for review/safe to merge, cc @inalgnu @samlesu @jmorel 

It might be easier to merge this PR into the `feature/generic-composite` branch first, and then review/test both composite and aggregate at the same time. Up to you. Feel free to do the merge!

--- 

I tested this in the following configuration and it's :+1: 

- substra: `dev` 
- substra-backend: `dev`
- substra-chaincode `aggregate`
- substra-tests: `master`

The only tests that fail also fail when I use the `dev` chaincode branch (the failing tests are all related to compute plans) 

Note that the only piece that is **not** implemented yet is permissions for aggregate traintuples (we currently disregard them completely).

File: chaincode/input_test.go [hunk 1]
```diff
@@ -206,6 +234,28 @@ func (traintuple *inputCompositeTraintuple) getArgs() [][]byte {
 	return args
 }
 
+func (traintuple *inputAggregateTuple) createDefault() [][]byte {
```
Comment (inelgnu) [inline chaincode/input_test.go]: shouldn't this receiver be called aggregatetuple?

File: chaincode/ledger_db.go [hunk 2]
```diff
@@ -289,6 +301,18 @@ func (db *LedgerDB) GetCompositeTraintuple(key string) (CompositeTraintuple, err
...
 }
 
+// GetAggregateTuple fetches a AggregateTuple from the ledger using its unique key
+func (db *LedgerDB) GetAggregateTuple(key string) (AggregateTuple, error) {
+	traintuple := AggregateTuple{}
+	if err := db.Get(key, &traintuple); err != nil {
+		return traintuple, err
+	}
+	if traintuple.AssetType != AggregateTupleType {
+		return traintuple, errors.NotFound("traintuple %s not found", key)
```
Comment (inelgnu) [inline chaincode/ledger_db.go]:
```suggestion
		return aggregatetuple, errors.NotFound("aggregatetuple %s not found", key)
```

**Review by inelgnu — APPROVED** [review]

**Comment by inelgnu** [comment]
Thank you @AurelienGasser

**Pull request merged**